prow/test/integration/fakeghserver/fakeghserver.go

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// fakeghserver serves github API for integration tests.
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/sirupsen/logrus"
+	prowgh "k8s.io/test-infra/prow/github"
+
+	"k8s.io/test-infra/prow/github/fakegithub"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+var ()
+
+type options struct {
+	port int
+}
+
+func (o *options) validate() error {
+	return nil
+}
+
+func flagOptions() *options {
+	o := &options{}
+	flag.IntVar(&o.port, "port", 8888, "Port to listen on.")
+	return o
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := flagOptions()
+	flag.Parse()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid arguments.")
+	}
+	defer interrupts.WaitForGracefulShutdown()
+	ghClient := fakegithub.NewFakeClient()
+
+	mux := http.NewServeMux()
+	// setup common handlers for local and deployed runs
+	mux.Handle("/", reposHandler(ghClient))
+	mux.Handle("/fakeghserver/", http.StripPrefix("/fakeghserver", reposHandler(ghClient)))
```

Comment:
Ah I see now. This is just to expose the same interface cluster-externally without using the root path externally (since that will be used by Deck later).

Another option would be to deploy testing services like `fakeghserver` to their own namespace with their own ingress on a different port. This would separate http routing for the testing services from the components under test and let us use a more standard ingress for the Prow components. (This wouldn't work with GKE ingresses, but should with ingress-nginx here.)

We will never need to serve a `/fakeghserver` path from Deck so the current solution should be fine, just food for thought.

Reply:
just standard name based virtual hosting should also work, no need for multiple ports

Reply:
Updated to use gorilla.mux to make this easier